[enhancement] kafkareceiver - permanent error handling option (on_permanent_error)

bohub12 opened the pull request:
#### Description

- Changed unmarshalling error to be treated as a permanent error
- Added an option in `kafkareceiver` to control message marking behavior based on permanent error occurrence
- Backward compatibility: if `on_permanent_error` is not specified, it inherits the value of `on_error`

#### Link to tracking issue
- Resolves : https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/41333

#### Testing

- Added unit tests
- Verified `on_permanent_error` behavior with both Sarama and Franz in e2e tests (producing messages to Kafka that trigger unmarshalling errors)

#### Documentation

- Updated kafkareceiver README

bohub12 committed:
[enhancement] kafkareceiver - on_permanent_error option - permanent error handling

axw commented on receiver/kafkareceiver/testdata/config.yaml:
```diff
@@ -60,4 +60,41 @@ kafka/rebalance_strategy:
+  message_marking:
+    after: true
+    on_error: true
+    on_permanent_error: true
```
Should we make this one `on_permanent_error: false`, to show that if the field is explicitly configured it will not inherit the value of `on_error`?

bohub12 replied:
I agree with what you said. I made the correction in the commit below.
- commit : https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41883/commits/67b9bad2a6f0685db3fcb27be6aa909731f7fc09

axw commented on receiver/kafkareceiver/kafka_receiver_test.go:
```diff
@@ -469,9 +470,21 @@ func TestReceiver_MessageMarking(t *testing.T) {
+			markAfter:           true,
+			markErrors:          true,
+			markPermanentErrors: false,
+			errorShouldRestart:  true,
```
```suggestion
			errorShouldRestart:  true, // error is permanent, so it isn't marked and will cause a restart
```

axw reviewed (commented):
Thanks @bohub12! Looks good, just a couple of small comments

bohub12 committed:
apply reveiw

- link : https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/41883\#discussion_r2268461321

bohub12 committed:
Update receiver/kafkareceiver/kafka_receiver_test.go

Co-authored-by: Andrew Wilkins <axwalk@gmail.com>

axw reviewed (approved):
Thanks very much @bohub12!
